File: DS/Linked LIst/CircularSinglyLinkedList.java
package dsa;

class Node {

    int data;
    Node next;
}

class Runner {

    public static void main(String[] args) {
        LinkedList list = new LinkedList();
        list.insert(10);
        list.insert(20);
        list.insert(30);
        list.insert(40);
        //list.insert(50);
        //list.insert(60);
        //list.insert(70);
        //list.insertAt(0,10);
        //list.insertAt(5,60);
        // list.insertAt(9,100);
        //list.delete();
        //list.deleteAt(4);
        //list.deleteAt(3);
        //list.deleteAt(6);
        list.reverseList();
        list.show();

        System.out.println("");

        System.out.println(list.getHead());
        System.out.println(list.getTail());
        // System.out.println(list.size());

    }
}

class LinkedList {

    Node head, tail;

    public void insert(int data) {
        Node newnode = new Node();
        newnode.next = null;
        newnode.data = data;

        if (head == null) {
            head = tail = newnode;
            newnode.next = head;
        } else {
            tail.next = newnode;
            tail = newnode;
            tail.next = head;
        }
    }

    public void insertAtBegin(int data) {
        Node newnode = new Node();
        newnode.data = data;

        newnode.next = head;
        head = newnode;
        tail.next = head;
    }

    public void insertAt(int index, int data) {
        if (index == 0) {
            insertAtBegin(data);
        } else {
            Node newnode = new Node();
            newnode.data = data;
            Node tempnode = head;
            for (int i = 0; i < index - 1; i++) {
                tempnode = tempnode.next;
            }

            if (index == size()) {
                tempnode.next = newnode;
                newnode.next = head;
                tail = tempnode.next;
                return;
            }
            newnode.next = tempnode.next;
            tempnode.next = newnode;
        }
    }

    public void delete() {
        head = head.next;
        tail.next = head;
    }

    public void deleteAt(int index) {
        if (head == null) {
            System.out.println("List is empty..");
            return;
        }
        if (index < 0 || index >= size()) {
            System.out.println("Invalid index..");
            return;
        }
        if (index == 0) {
            delete();
            return;
        }
        Node tempnode = head;
        for (int i = 0; i < index - 1; i++) {
            tempnode = tempnode.next;
        }
        if (index == size() - 1) {
            tempnode.next = head;
            tail = tempnode;
            tail.next = head;
            return;
        }
        tempnode.next = (tempnode.next).next;
    }

    public void reverseList() {

        tail = head;
        Node currnode = head, prevnode = head, nextnode = head;
        currnode = currnode.next;
        nextnode = nextnode.next.next;

        while (currnode != head) {
            currnode.next = prevnode;
            prevnode = currnode;
            currnode = nextnode;
            nextnode = nextnode.next;
        }
        head = prevnode;
        currnode.next = head;

    }

    public void show() {
        Node tempnode = head;
        do {
            System.out.print(tempnode.data + " ");
            tempnode = tempnode.next;
        } while (tempnode != head);
    }

    public int size() {
        if (head == null) {
            return 0;
        }
        int count = 1;
        Node tempnode = head;
        while (tempnode.next != head) {
            count++;
            tempnode = tempnode.next;
        }
        return count;
    }

    public int getHead() {
        return head.data;
    }

    public int getTail() {
        return tail.data;
    }
}

File: DS/Python/Linked List/CircularSinglyLinkedList.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None


class CircularSinglyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def insertAtBegin(self, data):
        new_node = Node(data)
        new_node.next = self.head
        if self.head is None:
            self.head = self.tail = new_node
        else:
            self.head = new_node
        self.tail.next = self.head

    def insertAtEnd(self, data):
        if self.head is None:
            self.insertAtBegin(data)
            return
        new_node = Node(data)
        new_node.next = self.head
        self.tail.next = new_node
        self.tail = new_node

    def insertAfter(self, item, data):
        if self.head:
            flag = False
            new_node = Node(data)
            temp = self.head
            while True:
                if temp.data == item:
                    flag = True
                    break
                temp = temp.next
                if temp == self.head:
                    break
            if flag:
                new_node.next = temp.next
                temp.next = new_node
                if new_node.next == self.head:
                    self.tail = new_node
            else:
                raise ValueError("Item not in the list!")
        else:
            raise ValueError("List is empty!")

    def remove(self, data):
        if self.head is None:
            raise ValueError("List is empty!")
        elif self.head.data == data:
            self.head = self.head.next
            self.tail.next = self.head
            return
        else:
            flag = False
            temp = self.head
            while True:
                if temp.data == data:
                    flag = True
                    break
                prev = temp
                temp = temp.next
                if temp == self.head:
                    break
            if flag:
                prev.next = temp.next
                if prev.next == self.head:
                    self.tail = prev
            else:
                raise ValueError("Item not in list!")

    def printList(self):
        if self.head:
            temp = self.head
            while True:
                print(temp.data, end=" ")
                temp = temp.next
                if temp == self.head:
                    break

    def size(self):
        count = 0
        if self.head:
            temp = self.head
            while True:
                count += 1
                temp = temp.next
                if temp == self.head:
                    break
        return count


csl = CircularSinglyLinkedList()

csl.insertAtEnd(50)
csl.insertAtEnd(60)
csl.insertAtBegin(40)
csl.insertAfter(40, 45)
csl.remove(60)
csl.insertAtEnd(60)

print("Head : ", csl.head.data)
print("Tail : ", csl.tail.data)

csl.printList()
